pythonforandroid/pythonpackage.py

```diff
@@ -0,0 +1,684 @@
+                       sys.real_prefix)
+
+
+def get_package_as_folder(dependency):
+    """ This function downloads the given package / dependency and extracts
+        the raw contents into a folder.
+
+        Afterwards, it returns a tuple with the type of distribution obtained,
+        and the temporary folder it extracted to.
+
+        Examples of returned values:
+
+        ("source", "/tmp/pythonpackage-venv-e84toiwjw")
+        ("wheel", "/tmp/pythonpackage-venv-85u78uj")
+
+        What the distribution type will be depends on what pip decides to
+        download.
+    """
+
+    venv_parent = tempfile.mkdtemp(
+        prefix="pythonpackage-venv-"
+    )
+    try:
+        # Create a venv to install into:
+        try:
+            if int(sys.version.partition(".")[0]) < 3:
+                # Python 2.x has no venv.
+                subprocess.check_output([
+                    sys.executable,  # no venv conflict possible,
+                                     # -> no need to use system python
+                    "-m", "virtualenv",
+                    "--python=" + _get_system_python_executable(),
+                    os.path.join(venv_parent, 'venv')
+                ], cwd=venv_parent)
+            else:
+                # On modern Python 3, use venv.
+                subprocess.check_output([
+                    _get_system_python_executable(), "-m", "venv",
+                    os.path.join(venv_parent, 'venv')
+                ], cwd=venv_parent)
+        except subprocess.CalledProcessError as e:
+            output = e.output.decode('utf-8', 'replace')
+            raise ValueError(
+                'venv creation unexpectedly ' +
+                'failed. error output: ' + str(output)
+            )
+        venv_path = os.path.join(venv_parent, "venv")
+
+        # Update pip and wheel in venv for latest feature support:
+        try:
+            filenotfounderror = FileNotFoundError
+        except NameError:  # Python 2.
+            filenotfounderror = OSError
+        try:
+            subprocess.check_output([
+                os.path.join(venv_path, "bin", "pip"),
+                "install", "-U", "pip", "wheel",
+            ])
+        except filenotfounderror:
+            raise RuntimeError(
+                "venv appears to be missing pip. "
+                "did we fail to use a proper system python??\n"
+                "system python path detected: {}\n"
+                "os.environ['PATH']: {}".format(
+                    _get_system_python_executable(),
+                    os.environ.get("PATH", "")
+                )
+            )
+
+        # Create download subfolder:
+        os.mkdir(os.path.join(venv_path, "download"))
+
+        # Write a requirements.txt with our package and download:
+        with open(os.path.join(venv_path, "requirements.txt"),
+                  "w", encoding="utf-8"
+                 ) as f:
+            def to_unicode(s):  # Needed for Python 2.
```

Comment:
Oh do we?

Reply:
well I didn't prefix all strings with `u` so in python 2 they are all bytes, but I use the python 3 `open()` since the python 2 `open()` as almost everything else in py 2 is kind of annoying to me :joy: so in that case this needs to be converted

or what exactly is your question?

Reply:
Yes this is pretty much my question, I just want to kill eventually not so necessary code branching

Reply:
It might be possible to avoid by littering `u` prefixes all over the code. (but conversions in other places might then pop up, I'm not sure, too much code to tell on a whim) But even if that worked, once we eventually drop python 2 we'd need to remove countless `u` prefixes instead of taking out that `to_unicode` inline function. That kind of makes more sense to me, but I'm happy to leave the decision to you

Reply:
(not that `u` in python 3 does any harm so I suppose we could just leave it in forever - but IMHO it is a little ugly given in python 3 it's just completely unnecessary and often just left away. Not sure what the more common approach is here usually...? But my intuition votes no `u` everywhere)

Reply:
Is this what `from __future__ import unicode_literals` is for?

Reply:
ah nice find! I just tested it though, and it seems pep517 and other external stuff will still pass in python-2-`str` even with that added, which would need conversion in a couple of places. I think in the end it might be less special handling to just have that `to_unicode` there